Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,588 @@
+-- the [cyclotomic](http://hackage.haskell.org/package/cyclotomic-0.5.1) package, one can use
+-- @[polarRat](https://hackage.haskell.org/package/cyclotomic-0.5.1/docs/Data-Complex-Cyclotomic.html#v:polarRat)
+-- 1 . @'fromRootOfUnity' to convert to a cyclotomic number.
+toComplex :: Floating a => RootOfUnity -> Complex a
+toComplex (RootOfUnity t)
+  | t == 1/2 = (-1) :+ 0
+  | t == 1/4 = 0 :+ 1
+  | t == 3/4 = 0 :+ (-1)
+  | otherwise = cis . (2*pi*) . fromRational $ t
+
+-- | For primes, define the canonical primitive root as the smallest such. For prime powers \(p^k\),
+-- either the smallest primitive root \(g\) mod \(p\) works, or \(g+p\) works.
+generator :: (Integral a, UniqueFactorisation a) => Prime a -> Word -> a
+generator p k
+  | k == 1 = modP
+  | otherwise = if powMod modP (p'-1) (p'*p') == 1 then modP + p' else modP
+  where p' = unPrime p
+        modP = case cyclicGroupFromFactors [(p,k)] of
+                 Just (Some cg) -> head $ filter (isPrimitiveRoot' cg) [2..p'-1]
+                 _ -> error "illegal"
+
+-- | Implement the function \(\lambda\) from page 5 of
+-- https://www2.eecs.berkeley.edu/Pubs/TechRpts/1984/CSD-84-186.pdf
+lambda :: Integer -> Int -> Integer
+lambda x e = ((powMod x (2*modulus) largeMod - 1) `shiftR` (e+1)) .&. (modulus - 1)
+  where modulus = bit (e-2)
+        largeMod = bit (2*e - 1)
+
+-- | For elements of the multiplicative group \((\mathbb{Z}/n\mathbb{Z})^*\), a Dirichlet
+-- character evaluates to a root of unity.
+eval :: DirichletCharacter n -> MultMod n -> RootOfUnity
+eval (Generated ds) m = foldMap (evalFactor m') ds
+  where m' = getVal $ multElement m
+
+-- | Evaluate each factor of the Dirichlet character.
+evalFactor :: Integer -> DirichletFactor -> RootOfUnity
+evalFactor m =
+  \case
+    OddPrime (toInteger . unPrime -> p) k (toInteger -> a) b ->
+      discreteLogarithmPP p k a (m `rem` p^k) `stimes` b
+    TwoPower k s b -> (if testBit m 1 then s else mempty)
+                   <> lambda m'' k `stimes` b
+                                       where m' = m .&. (bit k - 1)
+                                             m'' = if testBit m 1
+                                                      then bit k - m'
+                                                      else m'
+    Two -> mempty
+
+-- | A character can evaluate to a root of unity or zero: represented by @Nothing@.
+evalGeneral :: KnownNat n => DirichletCharacter n -> Mod n -> OrZero RootOfUnity
+evalGeneral chi t = case isMultElement t of
+                      Nothing -> Zero
+                      Just x -> NonZero $ eval chi x
+
+-- | Convert a Dirichlet character to a complex-valued function. As in `toComplex`, the result is
+-- inexact due to floating-point inaccuracies. See `toComplex`.
+toFunction :: (Integral a, RealFloat b, KnownNat n) => DirichletCharacter n -> a -> Complex b
+toFunction chi = asNumber toComplex . evalGeneral chi . fromIntegral
```

Comment:
Similar to discussion about `toRootOfUnity`, please make this function monomorphic in `a`, accepting `Mod n` only.

Reply:
Do you think it is worth keeping this function? I'd put it in originally for flexibility, but it is just a composition of other exposed functions, so perhaps it doesn't belong.

Reply:
Let's remove it then.